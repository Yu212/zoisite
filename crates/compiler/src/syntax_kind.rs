#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
#[repr(u16)]
pub enum SyntaxKind {
    Error,
    Whitespace,
    Eof,
    LineComment,
    String,
    Char,
    EqEq,
    Neq,
    Ge,
    Le,
    Gt,
    Lt,
    And,
    Or,
    Comma,
    Dot,
    Semicolon,
    Question,
    Colon,
    Ident,
    TypedIdent,
    TypeSpec,
    IdentTypeSpec,
    ArrayTypeSpec,
    OptionTypeSpec,
    TupleTypeSpec,
    LetKw,
    IfKw,
    ElseKw,
    WhileKw,
    BreakKw,
    ContinueKw,
    FunKw,
    TrueKw,
    FalseKw,
    NoneKw,
    Equals,
    Plus,
    Minus,
    Star,
    Slash,
    Percent,
    Number,
    OpenParen,
    CloseParen,
    OpenBracket,
    CloseBracket,
    OpenBrace,
    CloseBrace,
    Root,
    FuncDef,
    ParamList,
    NoneLiteral,
    IntLiteral,
    FloatLiteral,
    BoolLiteral,
    StringLiteral,
    CharLiteral,
    ArrayLiteral,
    BinaryExpr,
    PrefixExpr,
    ParenExpr,
    TupleExpr,
    RefExpr,
    IfExpr,
    FnCallExpr,
    IndexExpr,
    BlockExpr,
    Stmt,
    EmptyStmt,
    LetStmt,
    WhileStmt,
    BreakStmt,
    ContinueStmt,
    ExprStmt,
}

impl SyntaxKind {
    pub fn is_eof(self) -> bool {
        self == Self::Eof
    }

    pub fn is_error(self) -> bool {
        self == Self::Error
    }

    pub fn is_trivia(self) -> bool {
        matches!(self, SyntaxKind::Whitespace | SyntaxKind::LineComment)
    }
}

impl From<SyntaxKind> for rowan::SyntaxKind {
    fn from(kind: SyntaxKind) -> Self {
        Self(kind as u16)
    }
}
