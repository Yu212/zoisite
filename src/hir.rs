use ecow::EcoString;
use la_arena::Idx;
use rowan::TextRange;

use crate::resolve_context::FuncInfo;
use crate::scope::{FnId, VarId};

pub type ExprIdx = Idx<Expr>;
pub type StmtIdx = Idx<Stmt>;
pub type FuncIdx = Idx<Func>;

#[derive(Debug, Clone)]
pub struct Root {
    pub stmts: Vec<StmtIdx>,
    pub range: TextRange,
}

#[derive(Debug, Clone)]
pub struct Func {
    pub fn_info: Option<FuncInfo>,
    pub block: ExprIdx,
    pub range: TextRange,
}

#[derive(Debug, Clone)]
pub enum Stmt {
    EmptyStmt {
        range: TextRange,
    },
    LetStmt {
        var_id: Option<VarId>,
        expr: ExprIdx,
        range: TextRange,
    },
    WhileStmt {
        cond: ExprIdx,
        block: ExprIdx,
        range: TextRange,
    },
    BreakStmt {
        range: TextRange,
    },
    ContinueStmt {
        range: TextRange,
    },
    ExprStmt {
        expr: ExprIdx,
        range: TextRange,
    },
    FuncDef {
        func: FuncIdx,
        range: TextRange,
    }
}

#[derive(Debug, Clone)]
pub enum Expr {
    Missing,
    Binary {
        op: BinaryOp,
        lhs: ExprIdx,
        rhs: ExprIdx,
        range: TextRange,
    },
    Unary {
        op: UnaryOp,
        expr: ExprIdx,
        range: TextRange,
    },
    Ref {
        var_id: Option<VarId>,
        range: TextRange,
    },
    Tuple {
        elements: Vec<ExprIdx>,
        range: TextRange,
    },
    If {
        cond: ExprIdx,
        then_expr: ExprIdx,
        else_expr: Option<ExprIdx>,
        range: TextRange,
    },
    FnCall {
        fn_id: Option<FnId>,
        args: Vec<ExprIdx>,
        range: TextRange,
    },
    Index {
        main_expr: ExprIdx,
        index_expr: ExprIdx,
        range: TextRange,
    },
    Block {
        stmts: Vec<StmtIdx>,
        range: TextRange,
    },
    NumberLiteral {
        n: Option<u64>,
        range: TextRange,
    },
    BoolLiteral {
        val: bool,
        range: TextRange,
    },
    StringLiteral {
        val: Option<EcoString>,
        range: TextRange,
    },
    ArrayLiteral {
        len: Vec<ExprIdx>,
        initial: ExprIdx,
        range: TextRange,
    },
}

#[derive(Debug)]
pub struct Identifier {
    pub name: EcoString,
    pub range: TextRange,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum OpKind {
    BinaryOp(BinaryOp),
    UnaryOp(UnaryOp),
    PostfixOp(PostfixOp),
}

impl OpKind {
    pub fn binding_power(&self) -> (i8, i8) {
        match self {
            OpKind::BinaryOp(BinaryOp::Add) => (7, 8),
            OpKind::BinaryOp(BinaryOp::Sub) => (7, 8),
            OpKind::BinaryOp(BinaryOp::Mul) => (9, 10),
            OpKind::BinaryOp(BinaryOp::Div) => (9, 10),
            OpKind::BinaryOp(BinaryOp::Rem) => (9, 10),
            OpKind::BinaryOp(BinaryOp::Assign) => (2, 1),
            OpKind::BinaryOp(BinaryOp::EqEq) => (5, 6),
            OpKind::BinaryOp(BinaryOp::Neq) => (5, 6),
            OpKind::BinaryOp(BinaryOp::Ge) => (5, 6),
            OpKind::BinaryOp(BinaryOp::Le) => (5, 6),
            OpKind::BinaryOp(BinaryOp::Gt) => (5, 6),
            OpKind::BinaryOp(BinaryOp::Lt) => (5, 6),
            OpKind::BinaryOp(BinaryOp::And) => (3, 4),
            OpKind::BinaryOp(BinaryOp::Or) => (3, 4),
            OpKind::PostfixOp(PostfixOp::Index) => (12, -1),
            OpKind::UnaryOp(UnaryOp::Neg) => (-1, 11),
        }
    }
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum BinaryOp {
    Add,
    Sub,
    Mul,
    Div,
    Rem,
    Assign,
    EqEq,
    Neq,
    Ge,
    Le,
    Gt,
    Lt,
    And,
    Or,
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum PostfixOp {
    Index,
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub enum UnaryOp {
    Neg,
}
