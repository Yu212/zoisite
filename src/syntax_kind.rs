#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
#[repr(u16)]
pub enum SyntaxKind {
    Error,
    Whitespace,
    Eof,
    LineComment,
    String,
    EqEq,
    Neq,
    Ge,
    Le,
    Gt,
    Lt,
    Comma,
    Semicolon,
    Question,
    Colon,
    Ident,
    TypedIdent,
    TypeSpec,
    IdentTypeSpec,
    ArrayTypeSpec,
    OptionTypeSpec,
    TupleTypeSpec,
    LetKw,
    IfKw,
    ElseKw,
    WhileKw,
    BreakKw,
    ContinueKw,
    FunKw,
    TrueKw,
    FalseKw,
    Equals,
    Plus,
    Minus,
    Star,
    Slash,
    Percent,
    Number,
    OpenParen,
    CloseParen,
    OpenBracket,
    CloseBracket,
    OpenBrace,
    CloseBrace,
    Root,
    FuncDef,
    ParamList,
    NumberLiteral,
    BoolLiteral,
    StringLiteral,
    ArrayLiteral,
    BinaryExpr,
    PrefixExpr,
    ParenExpr,
    TupleExpr,
    RefExpr,
    IfExpr,
    FnCallExpr,
    IndexExpr,
    BlockExpr,
    Stmt,
    EmptyStmt,
    LetStmt,
    WhileStmt,
    BreakStmt,
    ContinueStmt,
    ExprStmt,
}

impl SyntaxKind {
    pub fn is_eof(self) -> bool {
        self == Self::Eof
    }

    pub fn is_error(self) -> bool {
        self == Self::Error
    }

    pub fn is_trivia(self) -> bool {
        matches!(self, SyntaxKind::Whitespace | SyntaxKind::LineComment)
    }
}

impl From<SyntaxKind> for rowan::SyntaxKind {
    fn from(kind: SyntaxKind) -> Self {
        Self(kind as u16)
    }
}
