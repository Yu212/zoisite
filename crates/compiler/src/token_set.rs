use crate::syntax_kind::SyntaxKind;

#[derive(Clone, Copy)]
pub struct TokenSet(u128);

impl TokenSet {
    pub const EMPTY: TokenSet = TokenSet(0);
    pub const ALL: TokenSet = TokenSet(!0);

    pub const fn new(kinds: &[SyntaxKind]) -> Self {
        let mut mask = 0_u128;
        let mut i = 0;
        while i < kinds.len() {
            mask |= 1_u128 << (kinds[i] as u16);
            i += 1;
        }
        TokenSet(mask)
    }

    pub const fn union(self, other: Self) -> Self {
        Self(self.0 | other.0)
    }

    pub const fn contains(self, kind: SyntaxKind) -> bool {
        self.0 >> (kind as u16) & 1 == 1
    }
}
