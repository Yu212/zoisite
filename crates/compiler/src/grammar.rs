use crate::hir::{BinaryOp, OpKind, PostfixOp, UnaryOp};
use crate::parser::{CompletedMarker, Parser};
use crate::syntax_kind::SyntaxKind;
use crate::token_set::TokenSet;

const RECOVERY_SET: TokenSet = TokenSet::new(&[SyntaxKind::Semicolon]);
const BOOL_SET: TokenSet = TokenSet::new(&[SyntaxKind::TrueKw, SyntaxKind::FalseKw]);

pub fn root(p: &mut Parser<'_>) {
    let m = p.start();
    p.eat_trivia();
    while !p.at(SyntaxKind::Eof) {
        let (_, semicolon) = stmt(p);
        if semicolon {
            p.expect(SyntaxKind::Semicolon);
        }
    }
    m.complete(p, SyntaxKind::Root);
}

pub fn stmt(p: &mut Parser<'_>) -> (CompletedMarker, bool) {
    match p.current() {
        SyntaxKind::Semicolon => (empty_stmt(p), true),
        SyntaxKind::LetKw => (let_stmt(p), true),
        SyntaxKind::WhileKw => (while_stmt(p), false),
        SyntaxKind::BreakKw => (break_stmt(p), true),
        SyntaxKind::ContinueKw => (continue_stmt(p), true),
        SyntaxKind::FunKw => (func_stmt(p), false),
        _ => expr_stmt(p),
    }
}

pub fn empty_stmt(p: &mut Parser<'_>) -> CompletedMarker {
    let m = p.start();
    m.complete(p, SyntaxKind::EmptyStmt)
}

pub fn func_stmt(p: &mut Parser<'_>) -> CompletedMarker {
    assert!(p.at(SyntaxKind::FunKw));
    let m = p.start();
    p.bump();
    p.expect(SyntaxKind::Ident);
    param_list(p);
    p.expect(SyntaxKind::Colon);
    type_spec(p);
    block_expr(p);
    m.complete(p, SyntaxKind::FuncDef)
}

pub fn while_stmt(p: &mut Parser<'_>) -> CompletedMarker {
    assert!(p.at(SyntaxKind::WhileKw));
    let m = p.start();
    p.bump();
    p.expect(SyntaxKind::OpenParen);
    expr(p, 0);
    p.expect(SyntaxKind::CloseParen);
    expr(p, 0);
    m.complete(p, SyntaxKind::WhileStmt)
}

pub fn break_stmt(p: &mut Parser<'_>) -> CompletedMarker {
    assert!(p.at(SyntaxKind::BreakKw));
    let m = p.start();
    p.bump();
    m.complete(p, SyntaxKind::BreakStmt)
}

pub fn continue_stmt(p: &mut Parser<'_>) -> CompletedMarker {
    assert!(p.at(SyntaxKind::ContinueKw));
    let m = p.start();
    p.bump();
    m.complete(p, SyntaxKind::ContinueStmt)
}

pub fn expr_stmt(p: &mut Parser<'_>) -> (CompletedMarker, bool) {
    let m = p.start();
    let res = expr(p, 0);
    let ends_with_block = res.map_or(false, |r| r.1);
    (m.complete(p, SyntaxKind::ExprStmt), !ends_with_block)
}

pub fn typed_ident(p: &mut Parser<'_>, optional: bool) -> CompletedMarker {
    let m = p.start();
    p.expect(SyntaxKind::Ident);
    if optional {
        if p.eat(SyntaxKind::Colon) {
            type_spec(p);
        }
    } else {
        p.expect(SyntaxKind::Colon);
        type_spec(p);
    }
    m.complete(p, SyntaxKind::TypedIdent)
}

pub fn type_spec(p: &mut Parser<'_>) -> Option<CompletedMarker> {
    match p.current() {
        SyntaxKind::OpenParen => {
            let m = p.start();
            p.bump();
            type_spec(p);
            p.expect(SyntaxKind::Comma);
            type_spec(p);
            while p.eat(SyntaxKind::Comma) {
                type_spec(p);
            }
            p.expect(SyntaxKind::CloseParen);
            Some(m.complete(p, SyntaxKind::TupleTypeSpec))
        },
        SyntaxKind::Ident => {
            let m = p.start();
            p.expect(SyntaxKind::Ident);
            let mut c = m.complete(p, SyntaxKind::IdentTypeSpec);
            loop {
                match p.current() {
                    SyntaxKind::OpenBracket => {
                        let m = c.precede(p);
                        p.bump();
                        p.expect(SyntaxKind::CloseBracket);
                        c = m.complete(p, SyntaxKind::ArrayTypeSpec);
                    },
                    SyntaxKind::Question => {
                        let m = c.precede(p);
                        p.bump();
                        c = m.complete(p, SyntaxKind::OptionTypeSpec);
                    },
                    _ => break
                }
            }
            Some(c)
        },
        _ => {
            p.error_and_recover(&[SyntaxKind::OpenParen, SyntaxKind::Ident], &RECOVERY_SET);
            None
        }
    }
}

pub fn param_list(p: &mut Parser<'_>) -> CompletedMarker {
    let m = p.start();
    p.expect(SyntaxKind::OpenParen);
    if !p.at(SyntaxKind::CloseParen) {
        typed_ident(p, false);
        while p.eat(SyntaxKind::Comma) {
            typed_ident(p, false);
        }
    }
    p.expect(SyntaxKind::CloseParen);
    m.complete(p, SyntaxKind::ParamList)
}

pub fn let_stmt(p: &mut Parser<'_>) -> CompletedMarker {
    assert!(p.at(SyntaxKind::LetKw));
    let m = p.start();
    p.bump();
    typed_ident(p, true);
    p.expect(SyntaxKind::Equals);
    expr(p, 0);
    m.complete(p, SyntaxKind::LetStmt)
}

pub fn expr(p: &mut Parser<'_>, min_binding_power: i8) -> Option<(CompletedMarker, bool)> {
    let (mut lhs, mut ends_with_block) = lhs(p)?;
    loop {
        let op: OpKind = match p.current() {
            SyntaxKind::Plus => OpKind::BinaryOp(BinaryOp::Add),
            SyntaxKind::Minus => OpKind::BinaryOp(BinaryOp::Sub),
            SyntaxKind::Star => OpKind::BinaryOp(BinaryOp::Mul),
            SyntaxKind::Slash => OpKind::BinaryOp(BinaryOp::Div),
            SyntaxKind::Percent => OpKind::BinaryOp(BinaryOp::Rem),
            SyntaxKind::Equals => OpKind::BinaryOp(BinaryOp::Assign),
            SyntaxKind::EqEq => OpKind::BinaryOp(BinaryOp::EqEq),
            SyntaxKind::Neq => OpKind::BinaryOp(BinaryOp::Neq),
            SyntaxKind::Ge => OpKind::BinaryOp(BinaryOp::Ge),
            SyntaxKind::Le => OpKind::BinaryOp(BinaryOp::Le),
            SyntaxKind::Gt => OpKind::BinaryOp(BinaryOp::Gt),
            SyntaxKind::Lt => OpKind::BinaryOp(BinaryOp::Lt),
            SyntaxKind::And => OpKind::BinaryOp(BinaryOp::And),
            SyntaxKind::Or => OpKind::BinaryOp(BinaryOp::Or),
            SyntaxKind::OpenBracket => OpKind::PostfixOp(PostfixOp::Index),
            _ => break,
        };
        let (left_binding_power, right_binding_power) = op.binding_power();
        if left_binding_power < min_binding_power {
            break;
        }
        p.bump();
        ends_with_block = false;
        let m = lhs.precede(p);
        if matches!(op, OpKind::PostfixOp(PostfixOp::Index)) {
            expr(p, 0);
            p.expect(SyntaxKind::CloseBracket);
            lhs = m.complete(p, SyntaxKind::IndexExpr);
        } else {
            let rhs = expr(p, right_binding_power);
            lhs = m.complete(p, SyntaxKind::BinaryExpr);
            if rhs.is_none() {
                break;
            }
        }
    }
    Some((lhs, ends_with_block))
}

pub fn lhs(p: &mut Parser<'_>) -> Option<(CompletedMarker, bool)> {
    match p.current() {
        SyntaxKind::NoneKw => Some((none_literal(p), false)),
        SyntaxKind::Number => Some((number_literal(p), false)),
        SyntaxKind::TrueKw | SyntaxKind::FalseKw => Some((bool_literal(p), false)),
        SyntaxKind::String => Some((string_literal(p), false)),
        SyntaxKind::OpenBracket => Some((array_literal(p), false)),
        SyntaxKind::Minus => Some((prefix_expr(p), false)),
        SyntaxKind::OpenParen => Some((paren_expr(p), false)),
        SyntaxKind::OpenBrace => Some((block_expr(p), true)),
        SyntaxKind::Ident if p.nth_at(1, SyntaxKind::OpenParen) => Some((fn_call_expr(p), false)),
        SyntaxKind::Ident => Some((ref_expr(p), false)),
        SyntaxKind::IfKw => Some(if_expr(p)),
        _ => {
            p.error_and_recover(&[SyntaxKind::Number, SyntaxKind::String, SyntaxKind::TrueKw, SyntaxKind::FalseKw, SyntaxKind::OpenBracket, SyntaxKind::Minus, SyntaxKind::OpenParen, SyntaxKind::OpenBrace, SyntaxKind::Ident, SyntaxKind::IfKw], &RECOVERY_SET);
            None
        }
    }
}

pub fn prefix_expr(p: &mut Parser<'_>) -> CompletedMarker {
    assert!(p.at(SyntaxKind::Minus));
    let m = p.start();
    p.bump();
    let op = OpKind::UnaryOp(UnaryOp::Neg);
    let (_, right_binding_power) = op.binding_power();
    expr(p, right_binding_power);
    m.complete(p, SyntaxKind::PrefixExpr)
}

pub fn paren_expr(p: &mut Parser<'_>) -> CompletedMarker {
    assert!(p.at(SyntaxKind::OpenParen));
    let m = p.start();
    p.bump();
    expr(p, 0);
    let mut comma = false;
    while p.eat(SyntaxKind::Comma) {
        expr(p, 0);
        comma = true;
    }
    p.expect(SyntaxKind::CloseParen);
    if comma {
        m.complete(p, SyntaxKind::TupleExpr)
    } else {
        m.complete(p, SyntaxKind::ParenExpr)
    }
}

pub fn ref_expr(p: &mut Parser<'_>) -> CompletedMarker {
    assert!(p.at(SyntaxKind::Ident));
    let m = p.start();
    p.bump();
    m.complete(p, SyntaxKind::RefExpr)
}

pub fn if_expr(p: &mut Parser<'_>) -> (CompletedMarker, bool) {
    assert!(p.at(SyntaxKind::IfKw));
    let m = p.start();
    p.bump();
    p.expect(SyntaxKind::OpenParen);
    expr(p, 0);
    p.expect(SyntaxKind::CloseParen);
    let res = expr(p, 0);
    let mut ends_with_block = res.map_or(false, |r| r.1);
    if p.eat(SyntaxKind::ElseKw) {
        let res = expr(p, 0);
        ends_with_block = res.map_or(false, |r| r.1);
    }
    (m.complete(p, SyntaxKind::IfExpr), ends_with_block)
}

pub fn fn_call_expr(p: &mut Parser<'_>) -> CompletedMarker {
    assert!(p.at(SyntaxKind::Ident));
    let m = p.start();
    p.bump();
    p.expect(SyntaxKind::OpenParen);
    if !p.at(SyntaxKind::CloseParen) {
        expr(p, 0);
        while p.eat(SyntaxKind::Comma) {
            expr(p, 0);
        }
    }
    p.expect(SyntaxKind::CloseParen);
    m.complete(p, SyntaxKind::FnCallExpr)
}

pub fn block_expr(p: &mut Parser<'_>) -> CompletedMarker {
    let m = p.start();
    p.expect(SyntaxKind::OpenBrace);
    while !p.at(SyntaxKind::Eof) {
        if p.at(SyntaxKind::CloseBrace) {
            empty_stmt(p);
            break;
        }
        let (_, semicolon) = stmt(p);
        if p.at(SyntaxKind::CloseBrace) {
            break;
        }
        if semicolon {
            p.expect(SyntaxKind::Semicolon);
        }
    }
    p.expect(SyntaxKind::CloseBrace);
    m.complete(p, SyntaxKind::BlockExpr)
}

pub fn none_literal(p: &mut Parser<'_>) -> CompletedMarker {
    assert!(p.at(SyntaxKind::NoneKw));
    let m = p.start();
    p.bump();
    m.complete(p, SyntaxKind::NoneLiteral)
}

pub fn number_literal(p: &mut Parser<'_>) -> CompletedMarker {
    assert!(p.at(SyntaxKind::Number));
    let m = p.start();
    p.bump();
    if p.eat(SyntaxKind::Dot) {
        m.complete(p, SyntaxKind::FloatLiteral)
    } else {
        m.complete(p, SyntaxKind::IntLiteral)
    }
}

pub fn string_literal(p: &mut Parser<'_>) -> CompletedMarker {
    assert!(p.at(SyntaxKind::String));
    let m = p.start();
    p.bump();
    m.complete(p, SyntaxKind::StringLiteral)
}

pub fn bool_literal(p: &mut Parser<'_>) -> CompletedMarker {
    assert!(p.at_set(&BOOL_SET));
    let m = p.start();
    p.bump();
    m.complete(p, SyntaxKind::BoolLiteral)
}

pub fn array_literal(p: &mut Parser<'_>) -> CompletedMarker {
    assert!(p.at(SyntaxKind::OpenBracket));
    let m = p.start();
    p.bump();
    expr(p, 0);
    p.expect(SyntaxKind::Semicolon);
    expr(p, 0);
    while p.eat(SyntaxKind::Comma) {
        expr(p, 0);
    }
    p.expect(SyntaxKind::CloseBracket);
    m.complete(p, SyntaxKind::ArrayLiteral)
}

#[cfg(test)]
mod tests {
    use crate::diagnostic::Diagnostic;
    use crate::language::SyntaxNode;
    use crate::lexer::Lexer;
    use crate::parser::Parser;

    fn parse(text: &str) -> (SyntaxNode, Vec<Diagnostic>) {
        let lexer = Lexer::new(text);
        let (tokens, _) = lexer.tokenize();
        let parser = Parser::new(tokens);
        parser.parse()
    }

    #[test]
    fn multi_stmt() {
        insta::assert_debug_snapshot!(parse("1; 2; 3;"));
    }

    #[test]
    fn func_def() {
        insta::assert_debug_snapshot!(parse("fun f(n: int): int { 1 }"));
    }

    #[test]
    fn let_stmt() {
        insta::assert_debug_snapshot!(parse("let a: int = 1;"));
    }

    #[test]
    fn expr_stmt() {
        insta::assert_debug_snapshot!(parse("1 + 2 * 3;"));
    }

    #[test]
    fn unary() {
        insta::assert_debug_snapshot!(parse("-1 * -2;"));
    }

    #[test]
    fn paren() {
        insta::assert_debug_snapshot!(parse("(1 + 2) * 3;"));
    }

    #[test]
    fn let_array() {
        insta::assert_debug_snapshot!(parse("let a: int[][] = [[1; 2]; 3];"));
    }

    #[test]
    fn ref_expr() {
        insta::assert_debug_snapshot!(parse("a + bc;"));
    }

    #[test]
    fn fn_call_expr() {
        insta::assert_debug_snapshot!(parse("f(a, 1);"));
    }

    #[test]
    fn assign_expr() {
        insta::assert_debug_snapshot!(parse("a = b[0] = 1;"));
    }

    #[test]
    fn block_expr() {
        insta::assert_debug_snapshot!(parse("{1} + {2};"));
    }

    #[test]
    fn type_spec() {
        insta::assert_debug_snapshot!(parse("let a: (int[]?, str, (char?, unit)) = 0;"));
    }
}
