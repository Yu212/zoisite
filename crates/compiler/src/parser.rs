use drop_bomb::DropBomb;
use rowan::TextRange;

use crate::diagnostic::{Diagnostic, DiagnosticKind};
use crate::event;
use crate::event::Event;
use crate::grammar::root;
use crate::language::SyntaxNode;
use crate::syntax_kind::SyntaxKind;
use crate::token::Token;
use crate::token_set::TokenSet;

pub struct Parser<'a> {
    tokens: Vec<Token<'a>>,
    cursor: usize,
    events: Vec<Event>,
}

impl<'a> Parser<'a> {
    pub fn new(tokens: Vec<Token<'a>>) -> Self {
        Self {
            tokens,
            cursor: 0,
            events: Vec::new(),
        }
    }
    pub fn parse(mut self) -> (SyntaxNode, Vec<Diagnostic>) {
        root(&mut self);
        event::process(&mut self.events, self.tokens)
    }
    pub fn current(&self) -> SyntaxKind {
        self.nth(0)
    }
    pub fn nth(&self, n: usize) -> SyntaxKind {
        let mut temp_cursor = self.cursor;
        for _ in 0..n {
            if self.tokens[temp_cursor].kind.is_eof() {
                break;
            }
            temp_cursor += 1;
            while self.tokens[temp_cursor].kind.is_trivia() {
                temp_cursor += 1;
            }
        }
        self.tokens[temp_cursor].kind.clone()
    }
    fn current_range(&self) -> TextRange {
        self.tokens[self.cursor].range
    }
    pub fn at(&self, kind: SyntaxKind) -> bool {
        self.nth_at(0, kind)
    }
    pub fn at_set(&self, set: &TokenSet) -> bool {
        set.contains(self.current())
    }
    pub fn nth_at(&self, n: usize, kind: SyntaxKind) -> bool {
        self.nth(n) == kind
    }
    pub fn eat(&mut self, kind: SyntaxKind) -> bool {
        if self.at(kind) {
            self.bump();
            true
        } else {
            false
        }
    }
    pub fn assert(&mut self, kind: SyntaxKind) {
        assert!(self.at(kind));
        self.bump();
    }
    pub fn bump(&mut self) {
        if self.current() != SyntaxKind::Eof {
            self.consume_token();
            self.eat_trivia();
        }
    }
    pub fn expect(&mut self, kind: SyntaxKind) -> bool {
        if self.eat(kind) {
            true
        } else {
            self.error(&[kind]);
            false
        }
    }
    pub fn error(&mut self, expected: &[SyntaxKind]) {
        self.events.push(Event::Error(Diagnostic::new(DiagnosticKind::UnexpectedToken {
            expected: expected.to_vec(),
            actual: self.current(),
        }, self.current_range())));
    }
    pub fn error_and_recover(&mut self, expected: &[SyntaxKind], recovery: &TokenSet) {
        if self.at_set(recovery) {
            self.error(expected);
            return;
        }
        let m = self.start();
        self.error(expected);
        self.bump();
        m.complete(self, SyntaxKind::Error);
    }
    pub fn eat_trivia(&mut self) {
        while self.current().is_trivia() {
            self.consume_token();
        }
    }
    pub fn start(&mut self) -> Marker {
        let pos = self.events.len();
        self.events.push(Event::Placeholder);
        Marker {
            pos,
            bomb: DropBomb::new("Marker must be either completed or abandoned")
        }
    }
    fn consume_token(&mut self) {
        self.events.push(Event::Token(self.current()));
        self.cursor += 1;
    }
}

pub struct Marker {
    pos: usize,
    bomb: DropBomb,
}

impl Marker {
    pub fn complete(mut self, p: &mut Parser<'_>, kind: SyntaxKind) -> CompletedMarker {
        self.bomb.defuse();
        let event = &mut p.events[self.pos];
        *event = Event::StartNode(kind, None);
        p.events.push(Event::FinishNode);
        CompletedMarker {
            pos: self.pos
        }
    }
    pub fn abandon(mut self) {
        self.bomb.defuse();
    }
}

pub struct CompletedMarker {
    pos: usize,
}

impl CompletedMarker {
    pub fn precede(self, p: &mut Parser<'_>) -> Marker {
        let m = p.start();
        match p.events[self.pos] {
            Event::StartNode(_, ref mut forward_parent) => {
                *forward_parent = Some(m.pos);
            },
            _ => unreachable!(),
        }
        m
    }
}
