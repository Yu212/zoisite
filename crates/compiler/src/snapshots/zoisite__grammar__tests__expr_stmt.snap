---
source: src/grammar.rs
expression: "parse(\"1 + 2 * 3;\")"
---
(
    Root@0..10
      ExprStmt@0..9
        BinaryExpr@0..9
          NumberLiteral@0..2
            Number@0..1 "1"
            Whitespace@1..2 " "
          Plus@2..3 "+"
          Whitespace@3..4 " "
          BinaryExpr@4..9
            NumberLiteral@4..6
              Number@4..5 "2"
              Whitespace@5..6 " "
            Star@6..7 "*"
            Whitespace@7..8 " "
            NumberLiteral@8..9
              Number@8..9 "3"
      Semicolon@9..10 ";"
    ,
    [],
)
